### ISSUE TITLE
[Analysis] v0.1-P2-N, UoMoE-N, and UoMoE-P2-N achieved 20% mAP50-95 on AI-TOD-v2, claiming the best e2e N-tier detectors for tiny object detection

### ISSUE THREAD

**Issue by skywalker-lt** [Tencent/YOLO-Master#126]
## 🎯 AP leap of UoMoE-N (`UltraOptimizedMoE`) over baseline v0.1-N on tiny-object detection (AI-TOD-v2)

On AI-TOD-v2 test split, the UoMoE-N model outperforms v0.1-N by **+9.6 points AP**, and **+20.0 points AP50**!

- Refer to section 1.4 in `wiki/MoE_Modules_Explanation_EN.md` for the structure of `UltraOptimizedMoE` module. 
- The mechanism of P2 variants (v0.1-P2-N & UoMoE-P2-N) is explained in #98.

## 📊 Results (AI-TOD-v2 test split)
| model | params (M) | GFLOPs@800 | AP | AP50 | AP75 | AP_vt | AP_t | AP_s | AP_m | Weights | 
|---|---|---|---|---|---|---|---|---|---|---|
| v0.1-N | 7.52 | 14.65 | 11.0 | 27.0 | 6.7 | 3.6 | 10.9 | 14.6 | 19.1 | [Download](https://github.com/skywalker-lt/YOLO-Master/releases/download/v0.1.0/yolo-master-v01-n-aitodv2.pt) |
| **UoMoE-N** | 7.42 | 14.61 | **20.6** | **47.0** | **14.5** | 5.9 | **18.9** | **28.0** | **35.1** | [Download](https://github.com/skywalker-lt/YOLO-Master/releases/download/v0.1.1/yolo-master-uomoe-n-aitodv2.pt) |
| v0.1-P2-N | 7.60 | 20.95 | **21.4** | **48.6** | **15.0** | 7.8 | **21.1** | 27.2 | 32.7 | [Download](https://github.com/skywalker-lt/YOLO-Master/releases/download/v0.1.1/yolo-master-v01-p2-n-aitodv2.pt) |
| UoMoE-P2-N | 7.50 | 20.91 | 21.1 | 47.8 | 14.7 | **8.6** | 20.4 | **27.6** | **33.8** | [Download](https://github.com/skywalker-lt/YOLO-Master/releases/download/v0.1.1/yolo-master-uomoe-p2-n-aitodv2.pt) |
| YOLOv12-X | 59.3 | 185.4 | 16.1 | 33.5 | 13.4 | 5.4 | 17.0 | 21.4 | 28.5 | N/A |

**YOLOv12 performance referenced from [Dome-DETR](https://arxiv.org/pdf/2505.05741)'s experiments, which applied similar args/hyps as ours.**

## Train & Eval Platform
| OS | CPU | Memory | GPU | VRAM | Driver | Python | CUDA | PyTorch | 
|---|---|---|---|---|---|---|---|---|
| Ubuntu Server 22.04 LTS | Xeon 8568Y+ | 512GB | 2x B200 | 2x 180GB | `580.126.09` | `3.14.6` | `12.8` | `2.11.0+cu128` |

## Training Methodology
- Script: `scripts/reproduce/reproduce_aitodv2.py` from PR #117 
- Model config: [`yolo-master-UoMoE-P2-N_aitodv2.yaml`](https://github.com/user-attachments/files/30063962/yolo-master-UoMoE-P2-N_aitodv2.yaml), [`yolo-master-UoMoE-N_aitodv2.yaml`](https://github.com/user-attachments/files/30064028/yolo-master-UoMoE-N_aitodv2.yaml), [`yolo-master-v0.1-P2-N_aitodv2.yaml`](https://github.com/user-attachments/files/30064039/yolo-master-v0.1-P2-N_aitodv2.yaml)
- Command: 
```bash
# YOLO-Master-v0.1-P2-N
python scripts/reproduce/reproduce_aitodv2.py --model v0.1-P2-N  --imgsz 800 --batch 64 --epochs 300
# YOLO-Master-UoMoE-N
python scripts/reproduce/reproduce_aitodv2.py --model UoMoE-N --imgsz 800 --batch 64 --epochs 300 --no-sparse-eval
# YOLO-Master-UoMoE-P2-N
python scripts/reproduce/reproduce_aitodv2.py --model UoMoE-P2-N --imgsz 800 --batch 64 --epochs 300 --no-sparse-eval
```

## Visualization and Comparison against YOLOv12

<img width="1300" height="845" alt="Image" src="https://github.com/user-attachments/assets/52991fcb-b5ae-4f9a-8560-23a556de4163" />

**Based on the experiments above, v0.1-P2-N, UoMoE-N and UoMoE-P2-N are by far the only three real-time N-tier detectors which achieved ≥20% AP on the AI-TOD-v2 benchmark, outperforming YOLOv12-X by 5 points AP with approx. 8~11% of its FLOPS, and marking YOLO-Master's architectural advantages on tiny object detection tasks.**

## 🔎 Finding
I observed a significant mAP leap on tiny-object detection when swapping v0.1-N backbone's three `ModularRouterExpertMoE` blocks by `UltraOptimizedMoE` (backbone/head unchanged, `[out, num_experts, top_k]` args identical): on the AI-TOD-v2 test split (14,018 imgs / 376,121 anns, 8 classes, mean object ~12px; official AI-TOD COCO metrics with conf 0.001, IoU 0.6, maxDets 1500; imgsz 800, 300 epochs, `optimizer=auto`, `lora_r=0`, batch 64 base / 32 P2), `UoMoE-N` jumps **+20.0 AP50 / +9.5 AP** over `v0.1-N` at smaller params (7.42 vs 7.52 M) and iso-FLOPs (14.61 vs 14.65 GF), which reveals a pure architecture win. However, when testing UoMoE with an extra P2 head, this stronger base proven otherwise: the extra stride-4 P2 head becomes redundant: P2 lifts the weak MoE (v0.1-N) by **+21.6 AP50** (27.0→48.6) but the strong base by only **+0.8** (47.0→47.8), showing the two interventions target the same tiny-object bottleneck and are substitutive rather than additive, with P2's residual value confined entirely to the very-tiny (<8px) bin (AP_vt 5.9→8.6, +2.7, making `UoMoE-P2-N` the best AP_vt of all four while flat elsewhere). Practically, `UoMoE-N` (no P2, 14.6 GF) lands within 1.6 AP50 of `v0.1-P2-N` (21.0 GF), a P2-variant-level accuracy at ~30% fewer FLOPs and no extra head. 

## Issue noticed during eval
One measurement issue from upstream: `ultralytics.utils.torch_utils.get_flops` profiles an uninitialised `torch.empty` 32×32 input scaled ×625, which for input-dependent MoE triggers a different expert mix each call and makes identical runs report different GFLOPs (we saw one UoMoE eval swing from 10.7 to 14.7); all FLOPs above therefore forced to a deterministic full-res measure, which is truely reproducible.

**Comment by skywalker-lt** [comment]
<img width="1901" height="1340" alt="Image" src="https://github.com/user-attachments/assets/4e2bac4e-19d5-4f76-b23a-2282a623fbd5" />

This diagram shows how the P2 and UoMoE-N variants relate to the baseline v0.1-N and EsMoE-N
